federation/executor.go

```diff
@@ -222,6 +223,17 @@ func (pathTargets *pathSubqueryMetadata) extractKeys(node interface{}, path []Pa
 		if err := pathTargets.extractKeys(next, path[1:]); err != nil {
 			return fmt.Errorf("elem %s: %v", next, err)
 		}
+	case KindType:
+		typ, ok := obj["__typename"].(string)
+		if !ok {
+			return fmt.Errorf("does not have string key __typename")
+		}
+
+		if typ == step.Name {
```

Comment:
What if typ != step.Name?

Reply:
In that case we just want to not extract any keys, but we don't want to error. I think this is easier to illustrate with an example

Let's say our query is 
```
query Foo {
				everyone {
					... on Admin {
						id
						hiding
					}
					... on User {
						id
					}
				}
			}
```
We first call extract keys on the results from "everyone", and when we figure out what the next object is https://github.com/samsarahq/thunder/blob/7fb9b9a865aece0ffc5b73f9398fdf8310e84886/federation/executor.go#L219 it has both a union of both the user and admin type. 

We then look at every element in the slice https://github.com/samsarahq/thunder/blob/7fb9b9a865aece0ffc5b73f9398fdf8310e84886/federation/executor.go#L183 , and recursively call it for the user and admin type. 

When we get to the line https://github.com/samsarahq/thunder/blob/7fb9b9a865aece0ffc5b73f9398fdf8310e84886/federation/executor.go#L232 , we want to check which type it is before extracting keys because we can call it with the wrong type. 


I can absolutely see how this is confusing, so I moved this check to when we are parsing the slice such that we only call extract keys if the type name is correct.

Reply:
Ah, I see, thanks for the explanation! 

Could you please squash the commit into the previous one?

Reply:
updated!